depmap-compute/depmap_compute/context.py

```diff
@@ -28,14 +31,128 @@
+    ):
+        """
+        A `context` dict should have:
+            - a `dimension_type` such as "depmap_model" (which is not used in this evaluator)
+            - an `expr` such as { "==": [ { "var": "slice/lineage/1/label" }, "Breast" ] }
+        """
+        self.expr = _encode_dots_in_vars(context["expr"])
+        self.slice_query_vars = context.get("vars", {})
+
+        # Takes a slice query, returns a dictionary of slice values (indexed by ID)
+        self.get_slice_data = get_slice_data
+
+        # The cache is used so that slice values only need to be looked up once per context.
+        # - The keys in this dictionary are slice queries encoded as a tuple
+        #   (ex. `("Chronos_combined", "SOX10", "feature_label")`).
+        # - The values are each an entire dictionary of slice values (indexed by given ID)
+        self.cache = {}
+
+    def is_match(self, given_id: str):
+        """
+        This evaluates `expr` against a `given_id`. It returns
+        True/False depending on if `given_id` satifies the conditions of
+        the expression, including any variables ("var" subexpressions) which
+        are bound by using a magic dict (_JsonLogicVarLookup) that does lookups lazily.
+        """
+        dictionary_override = _JsonLogicVarLookup(
+            given_id, self.cache, self.get_slice_data, self.slice_query_vars,
+        )
+
+        try:
+            return jsonLogic(self.expr, dictionary_override)
+        except (TypeError, ValueError) as e:
+            print("Exception evaluating", self.expr, "against", given_id)
+            print(e)
+            return False
+
+
+class _JsonLogicVarLookup(dict):
+    """
+    Context expressions use `var` fields to load data on the fly. 
+        For example, the following expression might be used to exclude certain given IDs from a query:
+            - { "!in": [ { "var": "given_id" }, ["1", "2", "3"] ] }
+
+    In order to populate these with real values, the JsonLogic library 
+    wants to be passed a dictionary it can use to look up values by variable name. 
+
+    However, we need to inject our own special cases and
+    caching, so we override the dictionary class with special functionality.
+    We don't need to "perfectly" override it; just well enough to trick the JsonLogic library.
+    Interesting thread on overriding the Dict class:
+    https://stackoverflow.com/questions/3387691/how-to-perfectly-override-a-dict
+    """
+
+    def __init__(
+        self,
+        given_id: str,
+        cache: dict,
+        get_slice_data: Callable[[SliceQuery], pd.Series],
+        slice_query_vars: dict[str, dict[str, str]],
+    ):
+        self.given_id = given_id
+        self.get_slice_data = get_slice_data
+        # The cache is stored outside of this class so it can be reused.
+        self.cache = cache
+        self.slice_query_vars = slice_query_vars
+
+    def __getitem__(self, var_name: Union[str, dict]) -> Any:
```

Comment:
I like calling this `var_name` but it's always a string. Maybe you were confused because it acts a key for `slice_query_vars` which is a dict?

Reply:
Oh good point! Yes, I was originally thinking this would take the slice query itself, and I forgot to update the type when I changed it to take the var_name instead.